### ISSUE TITLE
air-gap env: k3s with external containerd is not preloading images

### ISSUE THREAD

**Issue by ntx-ben** [k3s-io/k3s#3100]
**Environmental Info:**
K3s Version: v1.20.4+k3s1

Node(s) CPU architecture, OS, and Version: 
Alpine 3.13

Cluster Configuration:
1 instance ok k3s as server

**Describe the bug:**
Running k3s as part of an air-gapped environment. k3s is pointing to an external instance of containerd.

As far as I understand from this code here: [containerd.go](https://github.com/k3s-io/k3s/blob/88dd6019419488b7efd0cd41d47fb29eb465c1a0/pkg/agent/containerd/containerd.go#L118)

Unless k3s uses the built-in containerd, the air-gapped images will not be preloaded from the `/var/lib/rancher/k3s/agent/images` folder, as they usually are otherwise.

**Steps To Reproduce:**
- Put air-gap images tarball in preload folder
- Install k3s pointing to external containerd (i.e. `--container-runtime-endpoint unix:///run/containerd/containerd.sock`)
- `ctr images ls` shows no images pre-loaded

**Expected behavior:**
air-gapped images are imported regardless of using the built-in or external containerd instance.

**Actual behavior:**
air-gapped images are not pre-loaded

Is this something k3s could eventually support?

**Comment by brandond** [comment]
If you have an external containerd, you should be able to just use `ctr` to load the images yourself?

**Comment by dweomer** [comment]
@ntx-ben if the `ctr` you are using is from the external containerd installation you will need to use the correct containerd namespace (the `ctr` that comes with k3s does this for you auto-magically), i.e.:
```bash
ctr -n k8s.io image ls
```

**Comment by ntx-ben** [comment]
Thanks for your replies guys.

I guess I wanted to ask, rather, would it be possible for k3s to pre-load images regardless of the fact that its running containerd as a sub-process, or connecting to a standalone version. I was assuming that k3s would do that.

I'm lazy. Yes, I am now pre-loading images with ctr myself, but at first I was under the impression k3s would be doing that since previously I was using the k3s "bundled" containerd.

Anyway you can close this ticket.

**Comment by dweomer** [comment]
@ntx-ben is it actually not? A `ctr image ls` on a clean installation will never be empty as k3s will go ahead and attempt to download images from the internet (unless you are air-gapped of course). The output that you share was likely not listing images from the correct containerd namespace, `k8s.io`. Do you have running pods? If so, please `kubectl describe` one from `kube-system`, such as coredns, and see if there is a message there that the image was already present on the system.

**Comment by ntx-ben** [comment]
@dweomer Yeah I found out I had to add the `k8s.io `namespace to `ctr`, as you mentioned.

However, k3s is still not pre-loading images from my air gap tarball located at `/var/lib/rancher/k3s/agent/images/` folder. I did a clean install of k3s, pointing to a standalone containerd, and I can see pods stuck in pending state because the images cannot be fetched and are not already present. (I am in a air gap env.)

Hence my question/request for k3s to pre-load images if containerd is run as a sub-process, or as a standalone instance.

**Comment by dweomer** [comment]
> @dweomer Yeah I found out I had to add the `k8s.io `namespace to `ctr`, as you mentioned.
> 
> 
> 
> However, k3s is still not pre-loading images from my air gap tarball located at `/var/lib/rancher/k3s/agent/images/` folder. I did a clean install of k3s, pointing to a standalone containerd, and I can see pods stuck in pending state because the images cannot be fetched and are not already present. (I am in a air gap env.)
> 
> 
> 
> Hence my question/request for k3s to pre-load images if containerd is run as a sub-process, or as a standalone instance.

Okay, that does seem like a gap.

**Comment by brandond** [comment]
@dweomer we currently call preloadImages at the end of containerd.Run(), so if we don't start the embedded containerd, images never get preloaded.

https://github.com/k3s-io/k3s/blob/cb73461a5bd69cc91f3dc5fec3ffbbd8e970132a/pkg/agent/containerd/containerd.go#L120

**Comment by dweomer** [comment]
@brandond Thanks, I hadn't poked in the code yet but that reasoning makes sense.
